README.md
Graffiticode Compiler Framework
---
An open source framework for writing [graffiticode](https://graffiticode.com) compilers.

## Packages
- [`compiler-framework`](./packages/compiler-framework)
  
  An application to run graffiticode compilers. The application handles several aspects of serving a language and provides sensible defaults for optional parts.

- __WIP__ [`create-compiler`](./packages/create-compiler)

  A CLI that scaffolds a new langugage

- __WIP__ [`deployer`](./packages/deployer)

  A CLI that deploys graffiticode packages

- __WIP__ [`tracing`](./packages/tracing)

  A module to setup opentelemetry tracing for graffiticode applications

# Contributing

Contributions to this library are welcome and encouraged. See
[CONTRIBUTING](CONTRIBUTING.md) for more information on how to get started.

__NOTE__: _This is not an officially supported Google product._


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.3.1](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.3.0...v0.3.1) (2023-01-08)

**Note:** Version bump only for package root





# [0.3.0](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.2.0...v0.3.0) (2022-05-24)


### Features

* upgraded tracing deps and sdk creation ([b753f85](https://github.com/KevinDyer/node-graffiticode-compiler-framework/commit/b753f854af9fb4ba2bfb9e0823ac886c876a8a6f))





# [0.2.0](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.15...v0.2.0) (2022-05-23)


### Bug Fixes

* updated firebase-admin to remove node-forge insecure dep ([5627724](https://github.com/KevinDyer/node-graffiticode-compiler-framework/commit/56277246c6829cc9eeba6b9856f3ad15692b0d7a))
* updated firebase-admin to remove node-forge insecure dep in example code ([2a8576e](https://github.com/KevinDyer/node-graffiticode-compiler-framework/commit/2a8576ea3a0a065d83e1e3e232b9a9e5a8d574e7))
* updated npm deps to fix insecure versions in base ([8040537](https://github.com/KevinDyer/node-graffiticode-compiler-framework/commit/80405370c4dc750194392b8fdae052a086b9e8f0))
* updated npm deps to fix insecure versions in create-compiler ([62fcc3f](https://github.com/KevinDyer/node-graffiticode-compiler-framework/commit/62fcc3f90f47eacdc15d66236663715fad9de3e2))


### Features

* add nodejs auto instrumentations ([9633954](https://github.com/KevinDyer/node-graffiticode-compiler-framework/commit/96339545051b24eba8a0f6693d2e085d7bacf21d))





## [0.1.15](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.14...v0.1.15) (2022-03-06)

**Note:** Version bump only for package root





## [0.1.14](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.13...v0.1.14) (2022-03-06)

**Note:** Version bump only for package root





## [0.1.13](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.12...v0.1.13) (2021-11-07)

**Note:** Version bump only for package root





## [0.1.12](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.11...v0.1.12) (2021-11-03)

**Note:** Version bump only for package root





## [0.1.11](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.10...v0.1.11) (2021-11-03)

**Note:** Version bump only for package root





## [0.1.10](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.9...v0.1.10) (2021-10-31)

**Note:** Version bump only for package root





## [0.1.9](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.8...v0.1.9) (2021-10-24)

**Note:** Version bump only for package root





## [0.1.8](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.7...v0.1.8) (2021-07-18)

**Note:** Version bump only for package root





## [0.1.7](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.6...v0.1.7) (2021-07-18)

**Note:** Version bump only for package root





## [0.1.6](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.5...v0.1.6) (2021-07-18)

**Note:** Version bump only for package root





## [0.1.5](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.4...v0.1.5) (2021-07-15)

**Note:** Version bump only for package root





## [0.1.4](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.3...v0.1.4) (2021-07-15)

**Note:** Version bump only for package root





## [0.1.3](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.2...v0.1.3) (2021-07-15)

**Note:** Version bump only for package root





## [0.1.2](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.1.1...v0.1.2) (2021-07-15)

**Note:** Version bump only for package root





## [0.1.1](https://github.com/KevinDyer/node-graffiticode-compiler-framework/compare/v0.0.18...v0.1.1) (2021-07-15)

**Note:** Version bump only for package root


test/invoker.test.ts
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as bent from 'bent';
import * as http from 'http';
import * as net from 'net';
import {
  getServer,
  Compiler,
  AuthError,
  InvalidArgumentError,
} from '../src/invoker';

class FakeCompiler implements Compiler {
  language: string;
  _authError?: AuthError;
  _validateError?: InvalidArgumentError;
  _compileError?: Error;
  assetPath?: string;

  constructor(language: string) {
    this.language = language;
  }

  setAuthError(err?: AuthError) {
    this._authError = err;
  }

  setValidateError(err?: InvalidArgumentError) {
    this._validateError = err;
  }

  setCompileError(err?: Error) {
    this._compileError = err;
  }

  setAssertPath(assetPath: string) {
    this.assetPath = assetPath;
  }

  // tslint:disable-next-line:no-any
  async compile(code: {}, data: {}, config: {}): Promise<any> {
    if (this._compileError) {
      throw this._compileError;
    }
    return 'foo';
  }

  // tslint:disable-next-line:no-any
  async validate(code: {}, data: {}, config: {}): Promise<any> {
    if (this._validateError) {
      throw this._validateError;
    }
  }

  // tslint:disable-next-line:no-any
  async auth(token: string): Promise<any> {
    if (this._authError) {
      throw this._authError;
    }
  }
}

function createAndStartServer(
  compiler: Compiler
): Promise<{
  server: http.Server;
}> {
  return new Promise(resolve => {
    const server = getServer(compiler);
    server.listen(0, () => {
      resolve({
        server,
      });
    });
  });
}

function closeServer(server: http.Server) {
  return new Promise(resolve => server.close(resolve));
}

describe('invoker', () => {
  let compiler: FakeCompiler;
  let server: http.Server;
  beforeEach(async () => {
    compiler = new FakeCompiler('LTest');
    const ret = await createAndStartServer(compiler);
    server = ret.server;
  });

  afterEach(async () => {
    if (server) {
      await closeServer(server);
    }
  });

  it('should return language greeting', async () => {
    // Arrange
    const { port } = server.address() as net.AddressInfo;
    const getRoot = bent(`http://127.0.0.1:${port}`, 'string');

    // Act
    const actual = (await getRoot('/')) as string;

    // Assert
    expect(actual).toBe('Hello, LTest!');
  });

  it('should return language string', async () => {
    // Arrange
    const { port } = server.address() as net.AddressInfo;
    const getLang = bent(`http://127.0.0.1:${port}`, 'string');

    // Act
    const actual = (await getLang('/lang')) as string;

    // Assert
    expect(actual).toBe('LTest');
  });

  it('should return 404 if no assetPath', async () => {
    // Arrange
    const { port } = server.address() as net.AddressInfo;
    const getAsset = bent(`http://127.0.0.1:${port}`, 404);

    // Act
    const stream = (await getAsset('/style.css')) as bent.FetchResponse;

    // Assert
    expect(stream.statusCode).toBe(404);
  });

  it('should return asset if assetPath and exists', async () => {
    // Arrange
    await closeServer(server);
    compiler.setAssertPath(__dirname + '/assets');
    const ret = await createAndStartServer(compiler);
    server = ret.server;
    const { port } = server.address() as net.AddressInfo;
    const getAsset = bent(`http://127.0.0.1:${port}`, 'string');

    // Act
    const actual = (await getAsset('/style.css')) as string;

    // Assert
    expect(actual).toBe(`.foo {\n  display: flex;\n}`);
  });

  it('should return 404 if assetPath, but asset does not exist', async () => {
    // Arrange
    await closeServer(server);
    compiler.setAssertPath(__dirname + '/assets');
    const ret = await createAndStartServer(compiler);
    server = ret.server;
    const { port } = server.address() as net.AddressInfo;
    const getAsset = bent(`http://127.0.0.1:${port}`, 404);

    // Act
    const stream = (await getAsset('/boo.js')) as bent.FetchResponse;

    // Assert
    expect(stream.statusCode).toBe(404);
  });

  it('should return compiled value', async () => {
    // Arrange
    const { port } = server.address() as net.AddressInfo;
    const postCompile = bent(`http://127.0.0.1:${port}`, 'POST', 'json');

    // Act
    const actual = await postCompile('/compile', { code: {}, data: {} });

    // Assert
    expect(actual).toBe('foo');
  });

  it('should return error if compile throws', async () => {
    // Arrange
    compiler.setCompileError(new Error('failed to compile'));
    const { port } = server.address() as net.AddressInfo;
    const postCompile = bent(`http://127.0.0.1:${port}`, 'POST', 500);

    // Act
    const stream = (await postCompile('/compile', {
      code: {},
      data: {},
    })) as bent.FetchResponse;
    const actual = await stream.text();

    // Assert
    expect(stream.statusCode).toBe(500);
    expect(actual).toBe('Internal Server Error');
  });

  it('should return error if auth throws', async () => {
    // Arrange
    compiler.setAuthError(new AuthError('invalid token'));
    const { port } = server.address() as net.AddressInfo;
    const postCompile = bent(`http://127.0.0.1:${port}`, 'POST', 401);

    // Act
    const stream = (await postCompile('/compile', {
      code: {},
      data: {},
    })) as bent.FetchResponse;
    const actual = await stream.text();

    // Assert
    expect(stream.statusCode).toBe(401);
    expect(actual).toBe('Unauthorized');
  });

  it('should return error if validate throws', async () => {
    // Arrange
    compiler.setValidateError(new InvalidArgumentError('invalid code'));
    const { port } = server.address() as net.AddressInfo;
    const postCompile = bent(`http://127.0.0.1:${port}`, 'POST', 400);

    // Act
    const stream = (await postCompile('/compile', {
      code: {},
      data: {},
    })) as bent.FetchResponse;
    const actual = await stream.text();

    // Assert
    expect(stream.statusCode).toBe(400);
    expect(actual).toBe('invalid code');
  });
});


src/auth/count.spec.ts
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {buildCount} from './count';

describe('count', () => {
  it('should call postAuth', async () => {
    // Arrange
    const postAuth = jest.fn().mockResolvedValue({});
    const count = buildCount({postAuth});
    const token = 'token';
    const lang = 'L0';
    const increment = 1;

    // Act
    await count({token, lang, increment});

    // Assert
    expect(postAuth).toHaveBeenCalledWith({
      path: '/count',
      data: {
        jwt: token,
        lang,
        count: increment,
      },
    });
  });
  it('should throw if increment is less than one', async () => {
    // Arrange
    const postAuth = jest.fn().mockResolvedValue({});
    const count = buildCount({postAuth});
    const token = 'token';
    const lang = 'L0';
    const increment = 0;

    // Act
    await expect(count({token, lang, increment})).rejects.toThrow(
      'increment must be greater than zero'
    );

    // Assert
  });
});


src/auth/post-auth.spec.ts
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {buildPostAuth} from './post-auth';

describe('postAuth', () => {
  it('should ...', async () => {
    // Arrange
    const authCall = jest.fn().mockResolvedValue({});
    const bent = jest.fn().mockReturnValue(authCall);
    const postAuth = buildPostAuth({
      bent,
      baseUrl: 'http://auth.example.com',
    });
    const path = '/count';
    const data = {
      jwt: 'jwt',
      lang: 'lang',
      count: 1,
    };

    // Act
    await postAuth({path, data});

    // Assert
  });
});


src/auth/count.ts
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {PostAuthFunction} from './post-auth';

export interface CountRequest {
  token: string;
  lang: string;
  increment: number;
}

export interface CountFunction {
  (req: CountRequest): Promise<void>;
}

export interface BuildCountRequest {
  postAuth: PostAuthFunction;
}

export function buildCount(buildReq: BuildCountRequest): CountFunction {
  return async function count(req: CountRequest): Promise<void> {
    if (req.increment < 1) {
      throw new Error('increment must be greater than zero');
    }
    await buildReq.postAuth({
      path: '/count',
      data: {
        jwt: req.token,
        lang: req.lang,
        count: req.increment,
      },
    });
  };
}


src/auth/index.ts
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as bent from 'bent';

import {buildCount} from './count';
import {buildPostAuth} from './post-auth';
import {buildCreateValidateToken} from './validate-token';

const baseUrl = 'https://auth.artcompiler.com';
const postAuth = buildPostAuth({bent, baseUrl});
const count = buildCount({postAuth});
const createValidateToken = buildCreateValidateToken({postAuth, count});

// Exports
export * from './count';
export * from './post-auth';
export * from './validate-token';
export {count, postAuth, createValidateToken};


src/auth/post-auth.ts
export interface ValidateTokenPayload {
  jwt: string;
  lang: string;
}

export interface CountPayload {
  jwt: string;
  lang: string;
  count: number;
}

export interface PostAuthRequest {
  path: string;
  data: ValidateTokenPayload | CountPayload;
}

export interface PostAuthResponse {
  address: string;
  access: string;
}

export interface PostAuthFunction {
  (req: PostAuthRequest): Promise<PostAuthResponse>;
}

export interface BuildPostAuthRequest {
  bent: Function;
  baseUrl: string;
}

export function buildPostAuth(
  buildReq: BuildPostAuthRequest
): PostAuthFunction {
  const authCall = buildReq.bent(buildReq.baseUrl, 'POST', 'json');
  return async function postAuth(
    req: PostAuthRequest
  ): Promise<PostAuthResponse> {
    return (await authCall(req.path, req.data)) as PostAuthResponse;
  };
}


examples/simple/index.js
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
const {
  AuthError,
  createCloudFunction,
  createLambda,
  createValidateToken,
} = require('../../');

const validateToken = createValidateToken({ lang: 'L2' });

exports.compiler = {
  language: 'L2',

  async compile(code, data, config) {
    return null;
  },

  async auth(token) {
    const res = await validateToken({ token });
    if (res.access.indexOf('compile') === -1) {
      throw new AuthError('User does not have compile access');
    }
  },

  assetPath: __dirname + '/assets',
};

// This function can be used to serve a lambda function
exports.lambdaHandler = createLambda(this.compiler);

// This function can be used to serve a lambda function
exports.cloudFunctionHandler = createCloudFunction(this.compiler);


src/auth/validate-token.ts
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { CountFunction } from './count';
import { PostAuthFunction } from './post-auth';

export interface ValidateTokenRequest {
  token: string | null;
}

export interface ValidateTokenResponse {
  address: string;
  access: string;
}

export interface ValidateTokenFunction {
  (req: ValidateTokenRequest): Promise<ValidateTokenResponse>;
}

export interface BuildValidateTokenRequest {
  lang: string;
  count: CountFunction;
  postAuth: PostAuthFunction;
}

export function buildValidateToken(
  buildReq: BuildValidateTokenRequest
): ValidateTokenFunction {
  const cache = new Map();
  return async function validateToken(
    req: ValidateTokenRequest
  ): Promise<ValidateTokenResponse> {
    if (!req.token) {
      return {
        address: 'guest',
        access: 'compile',
      };
    }
    if (!cache.has(req.token)) {
      const res = await buildReq.postAuth({
        path: '/validate',
        data: {
          jwt: req.token,
          lang: buildReq.lang,
        },
      });
      cache.set(req.token, res);
    }
    buildReq.count({
      token: req.token,
      lang: buildReq.lang,
      increment: 1,
    });
    return cache.get(req.token);
  };
}

export interface CreateValidateTokenRequest {
  lang: string;
}

export interface CreateValidateTokenFunction {
  (req: CreateValidateTokenRequest): ValidateTokenFunction;
}

export interface BuildCreateValidateTokenRequest {
  postAuth: PostAuthFunction;
  count: CountFunction;
}

export function buildCreateValidateToken(
  buildReq: BuildCreateValidateTokenRequest
): CreateValidateTokenFunction {
  return function createValidateToken(
    req: CreateValidateTokenRequest
  ): ValidateTokenFunction {
    return buildValidateToken({
      postAuth: buildReq.postAuth,
      count: buildReq.count,
      lang: req.lang,
    });
  };
}


examples/simple/assets/lexicon.js
window.gcexports.globalLexicon = {
  val: {tk: 1, name: 'VAL', cls: 'function', length: 2, arity: 2},
  key: {tk: 1, name: 'KEY', cls: 'function', length: 2, arity: 2},
  len: {tk: 1, name: 'LEN', cls: 'function', length: 1, arity: 1},
  concat: {tk: 1, name: 'CONCAT', cls: 'function', length: 1, arity: 1},
  add: {tk: 1, name: 'ADD', cls: 'function', length: 2, arity: 2},
  mul: {tk: 1, name: 'MUL', cls: 'function', length: 2, arity: 2},
  pow: {tk: 1, name: 'POW', cls: 'function', length: 2, arity: 2},
  style: {tk: 1, name: 'STYLE', cls: 'function', length: 2, arity: 2},
  map: {tk: 1, name: 'MAP', cls: 'function', length: 2, arity: 2},
  apply: {tk: 1, name: 'APPLY', cls: 'function', length: 2, arity: 2},
  in: {tk: 1, name: 'IN', cls: 'function', length: 0, arity: 0},
  arg: {tk: 1, name: 'ARG', cls: 'function', length: 1, arity: 1},
};


src/invoker.ts
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as awsServerlessExpress from 'aws-serverless-express';
import * as lambda from 'aws-lambda';
import * as express from 'express';
import * as functions from 'firebase-functions';
import * as admin from 'firebase-admin';
import * as http from 'http';
import * as morgan from 'morgan';

export interface CompileFunction {
  // tslint:disable-next-line:no-any
  (code: {}, data: {}, config: {}): Promise<any>;
}

export interface ValidateFunction {
  // tslint:disable-next-line:no-any
  (code: {}, data: {}, config: {}): Promise<any>;
}

export interface AuthFunction {
  // tslint:disable-next-line:no-any
  (token: string): Promise<any>;
}

export class AuthError extends Error {}
export class InvalidArgumentError extends Error {}

export interface Compiler {
  language: string;
  compile: CompileFunction;
  auth?: AuthFunction;
  validate?: ValidateFunction;
  assetPath?: string;
}

function getModulePath(codeLocation: string): string | null {
  let path: string | null = null;
  try {
    path = require.resolve(codeLocation);
  } catch (ex) {
    console.error(ex.stack);
    try {
      // TODO: Decide if we want to keep this fallback.
      path = require.resolve(codeLocation + '/function.js');
    } catch (ex) {
      console.error(ex.stack);
    }
  }
  return path;
}

function getModuleProperty(
  codeLocation: string,
  target: string
  // tslint:disable-next-line:no-any
): any | null {
  try {
    const modulePath = getModulePath(codeLocation);
    if (modulePath === null) {
      console.error('Provided code is not a loadable module.');
      return null;
    }

    const module = require(modulePath);
    const property = target.split('.').reduce((code, targetPart) => {
      if (typeof code === 'undefined') {
        return undefined;
      } else {
        return code[targetPart];
      }
    }, module);
    return property;
  } catch (ex) {
    let additionalHint: string;
    // TODO: this should be done based on ex.code rather than string matching.
    console.log(ex.code);
    if (ex.stack.includes('Cannot find module')) {
      additionalHint =
        'Did you list all required modules in the package.json ' +
        'dependencies?\n';
    } else {
      additionalHint = 'Is there a syntax error in your code?\n';
    }
    console.error(
      `Provided module can't be loaded.\n${additionalHint}` +
        `Detailed stack trace: ${ex.stack}`
    );
    return null;
  }
}

export function getCompiler(
  codeLocation: string,
  target: string
): Compiler | null {
  return getModuleProperty(codeLocation, target) as Compiler;
}

function makeRootHandler(language: string): express.RequestHandler {
  return (req: express.Request, res: express.Response) => {
    res.status(200).send(`Hello, ${language}!`);
  };
}

function makeLanguageHandler(language: string): express.RequestHandler {
  return (req: express.Request, res: express.Response) => {
    res.status(200).send(language);
  };
}

function makeCompileHandler(compiler: Compiler): express.RequestHandler {
  return async (
    req: express.Request,
    res: express.Response,
    next: express.NextFunction
  ) => {
    try {
      const { auth, code, data, config = {} } = req.body;
      if (compiler.auth) {
        await compiler.auth(auth);
      }
      if (compiler.validate) {
        await compiler.validate(code, data, config);
      }
      const result = await compiler.compile(code, data, config);
      res.status(200).json(result);
    } catch (err) {
      next(err);
    }
  };
}

function makeErrorHandler(): express.ErrorRequestHandler {
  return async (
    err: Error,
    req: express.Request,
    res: express.Response,
    next: express.NextFunction
  ) => {
    // console.log(`---- ---- ----`);
    // console.error(err.stack);
    // console.log(`---- ---- ----`);
    if (err instanceof AuthError) {
      res.sendStatus(401);
    } else if (err instanceof InvalidArgumentError) {
      res.status(400).send(err.message);
    } else {
      res.sendStatus(500);
    }
  };
}

export function createApp(compiler: Compiler): express.Application {
  const app = express();
  if (process.env.DEBUG === 'true') {
    app.use(morgan('short'));
  } else {
    app.use(morgan('combined'));
  }
  if (compiler.assetPath) {
    app.use(express.static(compiler.assetPath));
  }
  app.use(express.json({ limit: '50mb' }));
  app.get('/', makeRootHandler(compiler.language));
  app.get('/lang', makeLanguageHandler(compiler.language));
  app.post('/compile', makeCompileHandler(compiler));
  app.use(makeErrorHandler());
  return app;
}

export function createLambda(compiler: Compiler): lambda.Handler {
  const app = createApp(compiler);
  const server = awsServerlessExpress.createServer(app);
  return function handler(
    event: lambda.APIGatewayProxyEvent,
    context: lambda.Context
  ) {
    awsServerlessExpress.proxy(server, event, context);
  };
}

export function getServer(compiler: Compiler): http.Server {
  return http.createServer(createApp(compiler));
}

export function createCloudFunction(
  compiler: Compiler
): functions.HttpsFunction {
  admin.initializeApp();
  return functions.https.onRequest(createApp(compiler));
}
